homeassistant/components/sensor/enertalk.py

```diff
@@ -0,0 +1,426 @@
+
+        html_response = html_response.format(response_message)
+        return web.Response(text=html_response.format(html_response),
+                            content_type='text/html')
+
+
+class EnerTalkOauth2Client:
+    """EnerTalk OAuth 2 Client."""
+
+    AUTHORIZE_ENDPOINT = 'https://auth.enertalk.com'
+    API_ENDPOINT = 'https://api2.enertalk.com'
+
+    def __init__(self, client_id, client_secret, config_path):
+        """Initialize the OAuth 2 Client."""
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.config_path = config_path
+        self.token = {}
+
+    def load_token(self):
+        """Load the stored token."""
+        import os
+        if not os.path.isfile(self.config_path):
+            raise PlatformNotReady
+
+        self.token = load_json(self.config_path)
+
+    def authorize_token_url(self, redirect_url):
+        """Authorize code grant type of OAuth 2.0 URL."""
+        return '{}/authorization' \
+               '?response_type=code&client_id={}&redirect_uri={}'\
+            .format(self.AUTHORIZE_ENDPOINT, self.client_id, redirect_url)
+
+    def fetch_access_token(self, code):
+        """Fetch Access Token."""
+        payload = {'grant_type': 'authorization_code',
+                   'client_id': self.client_id,
+                   'client_secret': self.client_secret,
+                   'code': code}
+        self.fetch_token({}, payload)
+
+    def renew_refresh_token(self):
+        """Renew Refresh Token."""
+        import base64
+        basic_auth = base64.standard_b64encode(
+            '{}:{}'.format(self.client_id, self.client_secret).encode('utf-8'))
+        basic_auth = basic_auth.decode('utf-8')
+        headers = {'Authorization': 'Basic {}'.format(basic_auth)}
+        payload = {'grant_type': 'refresh_token',
+                   'refresh_token': self.token[ATTR_REFRESH_TOKEN]}
+        self.fetch_token(headers, payload)
+
+    def fetch_token(self, add_headers, payload):
+        """Token is imported through the OAuth 2.0 API."""
+        headers = {'Content-Type': 'application/json'}
+        headers.update(add_headers)
+        try:
+            response = requests.post(
+                '{}/token'.format(self.AUTHORIZE_ENDPOINT),
+                headers=headers, json=payload, timeout=10)
+            response.raise_for_status()
+            _LOGGER.debug('JSON Response: %s', response.json())
+            self.token = response.json()
+            save_json(self.config_path, self.token)
+        except Exception as ex:
+            _LOGGER.error(
+                'Failed to update EnerToken status from token. Error: %s',
+                ex)
+            raise
+
+    def request(self, url):
+        """Enertalk API request."""
+        headers = {'Authorization': 'Bearer {}'.format(
+            self.token[ATTR_ACCESS_TOKEN]), 'accept-version': '2.0.0'}
+        try:
+            response = requests.get(
```

Comment:
Integration specific I/O and data parsing should be moved to standalone library published on pypi.